### PR TITLE
camelCase -> snake_case

### PR DESCRIPTION
**Pull request by jaimergp**
Following PEP8 recommendations and using table of replacements as provided in #368.

Note that `examples/` is not affected in this PR. They will be "modernized" in a separate PR.

## Questions

- [x] Shall I apply `snake_case` to public attributes and module-level variables?
- [x] Shall I apply `snake_case` to unexposed variables?

## Notes

* This PR introduces breaking changes! Check replacements in #368.
    * Special attention: `getFreeEnergyDifferences` is now `compute_free_energy_differences`, for consistency with other methods and functions.
* `pymbar.bar` and `pymbar.exp` _modules_ have been merged into `pymbar.other_estimators`. 

Summary of changed names:

```
AndersonDarling                                    anderson_darling
BARzero                                            bar_zero
computeBAR                                         bar
computeBARzero                                     bar_zero
computeCovarianceOfSums                            compute_covariance_of_sums
computeEffectiveSampleNumber                       compute_effective_sample_number
computeEntropyAndEnthalpy                          compute_entropy_and_enthalpy
computeEXP                                         exp
computeExpectations                                compute_expectations
computeExpectationsInner                           compute_expectations_inner
computeEXPGauss                                    exp_gauss
computeMultipleExpectations                        compute_multiple_expectations
computeOverlap                                     compute_overlap
computePerturbedFreeEnergies                       compute_perturbed_free_energies
computePMF                                         compute_pmf
detectEquilibration                                detect_equilibration
detectEquilibration_binary_search                  detect_equilibration_binary_search
EXPGauss                                           exp_gauss
generateConfidenceIntervals                        generate_confidence_intervals
generatePMF                                        generate_pmf
GetAnalytical                                      get_analytical
getFreeEnergyDifferences                           compute_free_energy_differences
getInformationCriteria                             get_information_criteria
getKDE                                             get_kde
getMBAR                                            get_mbar
getPMF                                             get_pmf
getWeights                                         weights
integratedAutocorrelationTime                      integrated_autocorrelation_time
integratedAutocorrelationTimeMultiple              integrated_autocorrelation_time_multiple
normalizedFluctuationCorrelationFunction           normalized_fluctuation_correlation_function
normalizedFluctuationCorrelationFunctionMultiple   normalized_fluctuation_correlation_function_multiple
OrderReplicates                                    order_replicates
PrintResults                                       print_results
QQPlot                                             qq_plot
statisticalInefficiency                            statistical_inefficiency
statisticalInefficiency_fft                        statistical_inefficiency_fft
statisticalInefficiencyMultiple                    statistical_inefficiency_multiple
subsampleCorrelatedData                            subsample_correlated_data
```

### CODE REVIEW

**Comment by mrshirts** [comment]
If if is best practice to snake_case public attributes, module-level variables, and unexposed variables, now it is the time to do.

**Comment by mrshirts** [comment]
Should we change things like EXP_gauss, BAR_gauss to exp_gauss and bar_gauss to be more snake_casey?

**Comment by jaimergp** [comment]
All in for snake_casiness!

**Comment by jaimergp** [comment]
There's a subtle problem:

If we go all lowercase, function `BAR` in `pymbar.bar` becomes `bar`. If we import that one in `pymbar.__init__`, it shadows `pymbar.bar` (module) to become `pymbar.bar` (function). Basically we can't have a function named as a module if we are going to expose that at a package level.

So, possible solutions:
* Keep the function named `BAR`, but it will be inconsistent with `bar_zero`.
* Rename the _function_ as `bennett_acceptance_ratio`. This must be done for similar acronyms (MBAR, PMF, etc).
* Rename the _modules_ so the acronyms are expanded.

**Comment by Lnaden** [comment]
So PEP8 says the following:

> Note: When using acronyms in CapWords, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError.

But I think that's for Classes more generally. I think for the snake case, it would still be `bar` since they are functions. `MBAR` would still be a thing since its a class. But is it really so bad to have `from pymbar.bar import bar`, especially since we mask the `pymbar.bar` module with `__all__` anyways?

**Comment by jaimergp** [comment]
It can lead to an inconsistency (I think that's why the tests are failing), where the _module_ `pymbar.bar` is unreachable. 

You are right about classes, so this only affects `BAR/bar` and `EXP/exp`.

**Comment by Lnaden** [comment]
How frequently do we think people manually import the `bar` and `exp` modules?

**Comment by jaimergp** [comment]
Not often, but the thing is that I think we literally make all the symbols in `pymbar.bar.*` unreachable this way?

**Comment by jaimergp** [comment]
The only way we can get there is by importing `pymbar.bar` directly, but before importing the bare `pymbar`, maybe?

```
import pymbar.bar
import pymbar
```

**Comment by Lnaden** [comment]
Well the only thing in the `bar` module right now to import are the BAR, BARZero (aside from the other imports), there is the logger for that module too, but thats it.

**Comment by Lnaden** [comment]
> The only way we can get there is by importing pymbar.bar directly, but before importing the bare pymbar, maybe?

I'm against that, lets keep it simple, I think if there are limited imports anyways we should just rename the modules (or put them in a new module, `other_estimators`?) and then import the relevant methods.

**Comment by jaimergp** [comment]
I am slightly inclined to rename modules rather than functions too. So let's see, would this work?

* bar -> bennett_acceptance_ratio
* exp -> exponential_averaging
* mbar -> multistate_bennett_acceptance_ratio
* mbar_solvers -> multistate_bennett_acceptance_ratio_solvers or simply solvers
* pmf -> potentials_of_mean_force

**Comment by Lnaden** [comment]
I would argue in favor of keeping mbar, mbar_solvers, and pmf as is. Its what people mostly will want the mbar repo for and it really is only bar and exp which are the problems. It may be better to move those two into their own module if we are looking for consistency.

**Comment by mrshirts** [comment]
> It may be better to move those two into their own module if we are looking for consistency.

I'd favor this.   other_estimators works for me.

**Comment by jaimergp** [comment]
Ok, sounds fair! To be clear, what option are we agreeing on?

* Option A: Merging `pymbar/bar.py` and `pymbar/exp.py` into a new file called `pymbar/other_estimators.py` 
* Option B: Creating a subpackage `pymbar/other_estimators/` that will contain `bar.py` and `exp.py`.

**Comment by Lnaden** [comment]
My vote is A, I like to avoid Subpackages when possible

**Comment by mrshirts** [comment]
Ditto, A.

**Comment by jaimergp** [comment]
Ok, that seems to do it. Please review whenever you have time.

I have dropped a couple of `TODO`s here and there regarding docstrings, too.

File: pymbar/tests/test_exp.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 import numpy as np
 import pytest
-from pymbar import exp as pyexp
+from pymbar import other_estimators as other
```
Comment (mrshirts) [inline pymbar/tests/test_exp.py]: I would probably do this as `import other_estimators as estimators`, since bar is an estimator, not an `other`.  Will make usage clearer.

File: pymbar/tests/test_bar.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 import numpy as np
 import pytest
-from pymbar import bar as pybar
+from pymbar import other_estimators as other
```
Comment (mrshirts) [inline pymbar/tests/test_bar.py]: Same as below (other -> estimator)

File: pymbar/timeseries.py [hunk 3]
```diff
@@ -972,4 +972,4 @@ def detect_equilibration_binary_search(A_t, bs_nodes=10):
             start = time_grid[k - 1]
             end = time_grid[k + 1]
 
-    return (t, g, Neff_max)
+    return t, g, Neff_max
```
Comment (mrshirts) [inline pymbar/timeseries.py]: Remind me why this change? (Just so I know good formatting practices)

Comment (jaimergp) [reply pymbar/timeseries.py]: They are redundant in this case. You do need to add them for single-element tuple for extra clarity (see [here](https://www.python.org/dev/peps/pep-0008/#when-to-use-trailing-commas)), but for single expression statements like a `return`, they can (and IMO should) be omitted for n-element tuples.

Comment (jaimergp) [reply pymbar/timeseries.py]: There might be more, now that you bring this up.

Comment (jaimergp) [reply pymbar/timeseries.py]: Yep, four more. All of them in `timeseries`. Do you want me to replace those too?

Comment (mrshirts) [reply pymbar/timeseries.py]: Sure, go for it!

**Comment by mrshirts** [comment]
I am OK with this being merged - not sure @jaimergp was waiting on anything else to merge.
